CloudSQLReplicator/google/cloud/pontem/sql/replicator/util/gcp_api_util.py

```diff
@@ -27,6 +31,112 @@ def _get_user_agent():
+
+    Args:
+      project (str): Project ID
+      credentials (google.auth.Credentials): Credentials to authorize client
+
+    Returns:
+      Authorized sqladmin service with CloudSQL Replicator user agent.
+  """
+  headers = _get_user_agent_header()
+  httplib2.Http.request.__func__.func_defaults = ('GET', None, headers, 5, None)
+  default_credentials, _ = google.auth.default()
+  authorized_http = google_auth_httplib2.AuthorizedHttp(
+      credentials or default_credentials
+  )
+
+  service = discovery.build(
+      'sqladmin',
+      'v1beta4',
+      http=authorized_http
+  )
+
+  return service
+
+def create_cloudsql_instance(database_instance_body=None,
+                             project=None,
+                             credentials=None):
+  """Provisions a Cloud SQL instance.
+
+    Args:
+      database_instance_body(JSON): Creation options
+      project(str): Project ID
+      credentials (google.auth.Credentials): Credentials to authorize client
+
+    Returns:
+      JSON: response from sqladmin.instances().insert() call
+  """
+
+  default_credentials, default_project = google.auth.default()
+  default_database_intance_body = {
+    'project': default_project,
+    'name': 'cloudsql-db-{}'.format(uuid.uuid4()),
+    'settings': {
+      'tier': 'db-n1-standard-2'
+    }
+  }
+  service = build_sql_admin_service(credentials or default_credentials)
+  request = service.instances().insert(
+      project=project or default_project,
+      body=database_instance_body or default_database_intance_body
+  )
+  response = request.execute()
+
+  return response
+
+def import_sql_database(database_instance,
+                        import_file_uri,
+                        project=None,
+                        credentials=None):
+  """Import database from SQL import file.
+
+    Args:
+      database_instance(str): Database instance id
+      project(str): Project ID
+      credentials (google.auth.Credentials): Credentials to authorize client
+
+    Returns:
+      JSON: response from sqladmin.instances().insert() call
+  """
+  default_credentials, default_project = google.auth.default()
+  service = build_sql_admin_service(credentials or default_credentials)
+  instances_import_request_body = {
+    "importContext": {
+      "kind": "sql#importContext",
+      "fileType": 'SQL',
+      "uri": import_file_uri,
+    }
+  }
+
+  request = service.instances().import_(
+      project=project or default_project,
+      instance=database_instance,
+      body=instances_import_request_body
+  )
+  response = request.execute()
+
+  return response
+
+
+def is_sql_operation_done(operation, project=None, credentials=None):
+  """Returns True if a SQL operation is done."""
+  default_credentials, default_project = google.auth.default()
+  service = build_sql_admin_service(credentials or default_credentials)
+  request = service.operations().get(project=project, operation=operation)
```

Comment:
Should't you be checking whether to use project or default project in/before the call?

Reply:
Loving the (var or default_var) idiom! Hadn't thought about that

Reply:
I thought about doing that (checking for project/credentials).  Since mutable default parameters are discouraged, I felt that the additional call on rarely called functions would be okay.  I open to suggestions, though.

Reply:
I think that's fine but I think didn't relay my concern properly.
I'm seeing that you get the default_project but you never use this variable and you just pass along the project variable which may be None to service.operations().get().

I fear you may be overlooking the use of the default variable if you actually don't care about default_project you can use the following idiom for variables that you're not gonna use:

default_creds, _ = google.auth.default() also
_, default_project = google.auth.default()

Reply:
Ok, I understand now.  This was in the is_sql_operation_done.  I fixed that.